lib/libuzfs/libuzfs_attr.c

```diff
@@ -0,0 +1,598 @@
+    nvlist_t **nvl, sa_handle_t *sa_hdl)
+{
+	int xattr_sa_size;
+	int err = sa_size(sa_hdl, sa_tbl[UZFS_XATTR], &xattr_sa_size);
+	ASSERT(err != ENOENT);
+	if (err != 0) {
+		return (err);
+	}
+
+	char *xattr_sa_data = vmem_alloc(xattr_sa_size, KM_SLEEP);
+	err = sa_lookup(sa_hdl, sa_tbl[UZFS_XATTR],
+	    xattr_sa_data, xattr_sa_size);
+	if (err == 0) {
+		err = nvlist_unpack(xattr_sa_data,
+		    xattr_sa_size, nvl, KM_SLEEP);
+	}
+	vmem_free(xattr_sa_data, xattr_sa_size);
+
+	return (err);
+}
+
+/*
+ * after settinging new entry, if the size
+ * of nvlist is too big to be saved by sa, this function
+ * will remove the name from nvlist and return EFBIG
+ */
+static int
+libuzfs_kvattr_update_nvlist(nvlist_t *nvl, const char *name,
+    const char *value, size_t value_size, size_t *xattr_sa_size)
+{
+	int err = nvlist_add_byte_array(nvl, name,
+	    (uchar_t *)value, (uint_t)value_size);
+	if (err != 0) {
+		return (err);
+	}
+
+	err = nvlist_size(nvl, xattr_sa_size, NV_ENCODE_XDR);
+	if (err != 0) {
+		return (err);
+	}
+
+	if (*xattr_sa_size > DXATTR_MAX_SA_SIZE ||
+	    *xattr_sa_size > SA_ATTR_MAX_LEN) {
+		err = nvlist_remove(nvl, name, DATA_TYPE_BYTE_ARRAY);
+		ASSERT(err != ENOENT);
+		if (err == 0) {
+			err = nvlist_size(nvl, xattr_sa_size, NV_ENCODE_XDR);
+		}
+		if (err != 0) {
+			return (err);
+		}
+		return (EFBIG);
+	}
+
+	return (0);
+}
+
+int
+libuzfs_inode_set_kvattr(libuzfs_dataset_handle_t *dhp, uint64_t ino,
+    const char *name, const char *value, uint64_t size,
+    int flags, uint64_t *txg)
+{
+	// only support value size <= 8K
+	ASSERT3U(size, <=, UZFS_XATTR_MAXVALUELEN);
+
+	sa_handle_t *sa_hdl = NULL;
+	objset_t *os = dhp->os;
+	int err = sa_handle_get(os, ino, NULL, SA_HDL_PRIVATE, &sa_hdl);
+	if (err != 0) {
+		return (err);
+	}
+
+	nvlist_t *nvl;
+	sa_attr_type_t *sa_tbl = dhp->uzfs_attr_table;
+	err = libuzfs_get_nvlist_from_handle(sa_tbl, &nvl, sa_hdl);
+	if (err == ENOENT) {
+		err = nvlist_alloc(&nvl, NV_UNIQUE_NAME, KM_SLEEP);
+	}
+	if (err != 0) {
+		goto out_handle;
+	}
+
+	boolean_t existed_in_sa = nvlist_exists(nvl, name);
+	boolean_t sa_space_enough = B_TRUE;
+	size_t xattr_sa_size = 0;
+	err = libuzfs_kvattr_update_nvlist(nvl, name,
+	    value, size, &xattr_sa_size);
+	if (err == EFBIG) {
+		sa_space_enough = B_FALSE;
+	} else if (err != 0) {
+		goto out_nvl;
+	}
+	ASSERT(xattr_sa_size > 0);
+
+	char *xattr_sa_data = NULL;
+	if (existed_in_sa || sa_space_enough) {
+		xattr_sa_data = vmem_alloc(xattr_sa_size, KM_SLEEP);
+		err = nvlist_pack(nvl, &xattr_sa_data, &xattr_sa_size,
+		    NV_ENCODE_XDR, KM_SLEEP);
+		if (err != 0) {
+			goto out_free_sa_data;
+		}
+	}
+
+	boolean_t existed_in_zap = B_FALSE;
+	uint64_t xattr_zap_obj = DMU_NEW_OBJECT;
+	uint64_t zap_entries_count = 0;
+	if (!existed_in_sa || !sa_space_enough) {
+		err = sa_lookup(sa_hdl, sa_tbl[UZFS_ZXATTR],
+		    &xattr_zap_obj, sizeof (xattr_zap_obj));
+
+		if (err == 0 && sa_space_enough) {
```

Comment:
it would be more clear to replace sa_space_enough with !existed_in_sa (they are equal due to L#316), means if name does not exists in sa and it has xattr_zap_obj, then we continue to check if name exists in zap.

Reply:
we can't just replace sa_space_enough with !existed_in_sa, because only when !existed_in_sa and sa_space_enough do we need to check existed_in_zap. 
if !existed_in_sa and !sa_space_enough, we just need to call zap_update and don't need to care existed_in_zap.
by the way, according to L#316, !existed_in_sa could means !existed_in_sa and !sa_space_enough

Reply:
agree